### DISCUSSION TITLE
Plans for after 7.0.13 is released

### DISCUSSION THREAD

**Discussion by MikeNeilson** [opendcs/opendcs#714] [General]
... or Oct 4th, which ever ends up coming first.

So we have a few large changes that need to happen to make development a bit smoother, and in some cases hopefully less duplicated, going forward. So a heads up that given the title and first line above I'm going to start working in earnest on the following:

1. Setting up the gradle build. This will allow dependabot to help us keep deps up to date and frankly, remove a *huge* maintenance burden off my back for things already covered by convention
2. Automating artifact pushes to GitHub Container Registry, Maven Central, GitHub Packages as appropriate. (snapshots/docker image to github, RCs and finals to mevan central - I have this working with gradle in other opendcs repos
3. Tag based releases - we use this internally at HEC and it makes it a lot easier for work to happen in multiple repos.
  4. as part of this I'm going to create a 7.0 branch from master and rename master to main. As there is a small chance the next element may be a bit too disruptive and a critical fix is needed
5. Finalize and merge in (to the new main initially) https://github.com/opendcs/opendcs/pull/681
  6. including work appropriate to supporting: https://github.com/opendcs/rest_api/issues/191
  7. The existing Pair<Database,TimeSeriesDatabase> will be changed to `OpenDcsDatabase` which will container the pair and allow retrieval of the invidiual instances until such time as everything is appropriately moved to be through `OpenDcsDatabase`
  8. `OpenDcsDatabase` will include some of the design ideas I've pitched in other issues, discussions, and the wiki.

My goal is to get that done by the end of October, but hopefully at least before thanksgiving. @ktarbet has agreed (been coerced?) into a willingness to review that rather large PR that is #861.

The use of the DataSource, and connection pool, is critical to required parallel operations of compproc and the routing scheduler, as well as allowing direct use of the OpenDCS core DAOs and such in the rest_api. 

Somewhere in there JDBI may finally come up, but that's been a struggle with how spaghetti the "Decodes Database" (`Database` and the whole DbIo system) is throughout the code.

@opendcs/opendcs-core-devs @opendcs/opendcs-triage

**Comment by ktarbet** [comment]
I like this idea of having a 7.0 branch that can have patches as needed for stability, and the work described above will go into main; If I'm understanding this.

Yes, I'm willing to review code!  I'm motivated by minimizing duplicate database access code.  Optimizations such as bind variables, security fixes, special cases for specific databases,  will have a dedicated default location instead of duplicate locations.

I'll try to respond in a reasonable time to PR's.    I would ask for separate PR's as much as possible, such as moving to gradle as a separate PR - I understand that would also include dropping the installer.

**Reply by MikeNeilson** [reply]
Yep. the crazy data source thing is going to be somewhat large. As I'd prefer not to leave "main" in a broken state for new developers. But the majority of the work identified above is separate efforts that should be reasonably managable.

**Comment by agilmore2** [comment]
What's the install process going forward for 7.0.14+ and the zipfile distribution?

**Reply by MikeNeilson** [reply]
1. Download zip or tar
2. Unzip or tar -xf
3. ???
4. Profit.

Which can change this, but currently the zip/tar has "opendcs-<version>" as the root directory. I'm inclined to keep it so that most users avoid accidents as it's generally easy to tell the unzip commands to strip the first directory.

**Reply by MikeNeilson** [reply]
My having forgot to include a few items in the distribution not withstanding, it appears to have been otherwise working okay for our users. (and it pretty much what the CWMS TAR always except except we're not removing the .bat for "unix" as we don't want to force the choice of zip vs tar on any platform.

Also in more serious terms, you'll want to change your PATH to include the new bin directory or use a symlink for "active version" instead if that works better for you.

**Comment by agilmore2** [comment]
We almost always use a symlink in our server installs.
IIRC, old installs had both the shell and .bat scripts anyway.
Some Reclamation users enjoyed the "place shortcuts on desktop" feature, but they can live.

**Reply by MikeNeilson** [reply]
It only worked on windows and certain linux desktops (e.g. not ubuntu). So I think definitely something to leave to the user, as nice as it was to have.
